disk: add a SDHC card over SPI driver.

nzmichaelh opened the pull request:
Hi there.  I'd like some early feedback on the SDHC card over SPI driver in this patch.  This is part of #5457 

Features:

- Uses the SPI bus to communicate with the card
- Detects and safely rejects SDSC (<= 2 GiB) cards
- Uses the optional CRC support for data integrity
- Retries resumable errors like CRC failure or temporary IO failure
- Works well with ELMFAT
- When used on a device with a FIFO or DMA, achieves >= 310 KiB/s on a 4 MHz bus

Tested on a mix of SanDisk, Samsung, 4V, and ADATA cards from 4 GiB to 32 GiB.

Some open questions are:

- [ ] The SD card protocol does multiple commands in one transaction.  The acquire/release implementation in the new SPI API seems to be broken - it uses a semaphore so the second call causes the device to hang
- [x] How to bind the device to the right SPI port.  It currently assumes SPI_0 but it'd be great to configure it via DeviceTree
- [x] Where should the CRC routines live - here or under drviers/crc?  Putting them under drivers/crc will make it easier to add support for the SAMD21 CRC peripheral in the future.

nzmichaelh committed:
disk: add a SDHC card over SPI driver.

Features:

- Uses the SPI bus to communicate with the card
- Detects and safely rejects SDSC (<= 2 GiB) cards
- Uses the optional CRC support for data integrity
- Retries resumable errors like CRC failure or temporary IO failure
- Works well with ELMFAT
- When used on a device with a FIFO or DMA, achieves >= 310 KiB/s on a
  4 MHz bus

Tested on a mix of SanDisk, Samsung, 4V, and ADATA cards from 4 GiB to
32 GiB.

Signed-off-by: Michael Hope <mlhx@google.com>

superna9999 commented:
> How to bind the device to the right SPI port. It currently assumes SPI_0 but it'd be great to configure it via DeviceTree
@ydamigos posted this for STM32 in #5582

superna9999 commented:
I have a comment over this SDHC only support, is it too complex to handle SDSC, SDHC and MMC cards ? Seems better to handle all these, no ?

superna9999 commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+/* Asserts or deasserts chip select */
+static void sdhc_set_cs(struct sdhc_data *data, int value)
+{
+	gpio_write(data->cs, GPIO_ACCESS_BY_PIN, data->pin, value);
```
Please use gpio_pin_write()

nzmichaelh replied:
Done.

superna9999 commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+	struct sdhc_data *data = dev->driver_data;
+	int err;
+
+	data->status = DISK_STATUS_UNINIT;
```
Could this driver support removable SDCards aswell with support for Card Detec GPIO ?

nzmichaelh replied:
Could do, yip, but it'd probably need changes in the FS layer as well.  At the moment the sequence is `boot -> fs_init -> disk_init` but to support card detect we'd probably want `boot -> register GPIO interrupt; interrupt -> app -> disk_init -> fs_init`

OK if I do that in a follow up PR?

superna9999 replied:
I thin kit can wait for a follow up if the current architecture is simple enough to expand.

nzmichaelh commented:
@superna9999 re: MMC and SDSC support, I'd rather not if that's OK.  The changes aren't huge - a change in the initialisation sequence and byte based instead of sector based addressing - but they're hard to test and I don't think they'd get used.  I haven't seen a MMC card in ages, and the <= 2 GiB SDSC cards are unusual (4 of the ~205 of SD cards on digitech.ch) and more expensive than a 4 GiB or 8 GiB card.

superna9999 commented:
@nzmichaelh You are right about MMC since the last eMMC do not support the SPI interface anymore.
But maybe we could go further and have a MMC/SD/SDHC framework with a SDHC-SPI interface driver.
Some MCUs now have a dedicated MMC controller to drive cheap eMMCs, and ultimately MMC support should be added.

tbursztyka commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+static u8_t sdhc_crc7_finish(u8_t crc) { return (crc << 1) | 1; }
+
+static u16_t sdhc_crc16(u16_t seed, const u8_t *src, size_t len)
+{
```
there is already a generic implementation of crc16, see include/crc16.h

nzmichaelh replied:
Ah, that's more epic.  The one in include/crc16 bitwise and 10x slower than this bytewise version (see #5098) and would drop the transfer rate by 2.3x [1].  My plan was to land this PR first, then pick up #5098, and then add support in drivers/crc for the hardware CRC unit on some devices (which is 4x faster again on the SAMD21).  WDYT?

[1] https://juju.nz/michaelh/post/2017/sdhc_speed/

tbursztyka commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+static int sdhc_rx_bytes(struct sdhc_data *data, u8_t *buf, int len)
+{
+	struct spi_buf tx_bufs[] = {
+		{.buf = (u8_t *)sdhc_ones, .len = len},
```
hm
{
.buf = ...,
.len = ...
};

would be nicer.

nzmichaelh replied:
Done globally.

tbursztyka commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+#include <gpio.h>
+#include <misc/__assert.h>
+#include <misc/byteorder.h>
+#include <net/buf.h>
```
You don't need this.

tbursztyka commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,1001 @@
+
+int disk_access_ioctl(u8_t cmd, void *buf)
+{
+	struct device *dev = sdhc_get_device();
```
this disk access api seems rather limited: only 1 disk is possible to get...

nzmichaelh replied:
Agreed, the Zephyr API is limited.  Longer term, I think Zephyr should switch to a block device driver interface so you can support, say, config in NFFS on internal flash and logging on FAT on a SD card.

Anything you'd like me to do here?

nashif replied:
we have a issue dealing with this and someone looking at it.

tbursztyka reviewed (changes requested):
Try to be consistent with line breaks:

var = something();
if (var ...) {

no need of line breaks here (sometime you do it, sometime not).

Also, try to aerate the code where relevant by adding a line break (pile of ifs and returns sometime, could gain by being put a bit appart.)

nzmichaelh commented:
@tbursztyka re: line breaks, I've updated it so if the use follows the assignment then there's no break between.  I also went through and generally added breaks between statements.

nzmichaelh commented:
@superna9999 longer term, something like Linux where there's a protocol vs host driver split SGTM.  I don't have a device with a SD host or the need for this ATM though.

nzmichaelh committed:
lib: add crc7.

Used as a checksum on command messages when talking with MMC cards.

Implemented using the unwound bytewise implementation from
https://en.wikipedia.org/wiki/Computation_of_cyclic_redundancy_checks
which is a good mix of size and speed.

The API and naming matches lib/crc7.c in Linux.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
Hi @tbursztyka @dbkinder.  This is now in decent state and ready for review.

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+The system has been tested with cards from Samsung, SanDisk, and 4V
+with sizes from 2 GiB to 32 GiB in single partition mode.
+
+MMC and SDSC (<= 2 GiB) cards are not supported and will be ignored.
```
... and cards > 32GB are ignored too?

nzmichaelh replied:
Clarified that it should work for higher capacity but hasn't been tested.

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+
+MMC and SDSC (<= 2 GiB) cards are not supported and will be ignored.
+
+The system does not support inserting or removing cards, so the card
```
maybe add:

    The system does not support inserting or removing cards while the system is running.
    The cards must be present at boot and must not be removed.

nzmichaelh replied:
Done

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+    };
+
+adds SDHC card slot on `spi1`, uses `PA27` for chip select, and runs
+the SPI bus at 24 MHz once the SDHC card has been initialised.
```
``initialized`` (American spelling)

Are there any tips for using the device after it's been initialized?

nzmichaelh replied:
Added a short usage section.  Unfortunately there's no generic FS documentation or sample ATM :(

ydamigos commented on dts/bindings/mmc/mmc-spi-slot.yaml:
```diff
@@ -0,0 +1,25 @@
+    compatible:
+      constraint: "zephyr,mmc-spi-slot"
+
+    gpios:
```
I propose to follow the same approach as linux/Documentation/devicetree/bindings/spi/spi-bus.txt. We should add a property cs-gpios in spi.yaml and modify extract_dts_includes.py to support it.

If, for example, the SPI has 2 CS lines, and the cs-gpios property looks like this:

`cs-gpios = <&gpioA 12 0> <&gpioC 11 0>;`

we have the following mapping:
```
cs0 : &gpioA 12 0
cs1 : &gpioC 11 0
```
The 0 in spi-device@0 and property reg = <0> point to cs0.

nzmichaelh replied:
@ydamigos that sounds fine.  Who should I ask to do the dts generator changes?

ydamigos commented:
I tried it on olimexino_stm32 and it doesn't work:
`[usb/msc] [ERR] mass_storage_init: Storage init ERROR !!!! - Aborting USB init`

It fails to control the CS pin. I did the follow change and the CS worked but I still get the above error.
```
diff --git a/subsys/disk/disk_access_sdhc.c b/subsys/disk/disk_access_sdhc.c
index 1f2fc1bb7..b45f119bc 100644
--- a/subsys/disk/disk_access_sdhc.c
+++ b/subsys/disk/disk_access_sdhc.c
@@ -975,4 +975,4 @@ int disk_access_init(void)
 static struct sdhc_data sdhc_data_0;

 DEVICE_AND_API_INIT(sdhc_0, "sdhc_0", sdhc_init, &sdhc_data_0, NULL,
-                   POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+                   APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);

```

tbursztyka commented:
@ydamigos Sounds like CONFIG_KERNEL_INIT_PRIORITY_DEFAULT is the culprit (either gpio or spi are initialized after sdhc, thus the issue. So init prio  should be tweaked instead of the level)

@nzmichaelh maybe it would be worth adding a generic CONFIG_SDHC_INIT_PRIORITY set to relevant value then.

ydamigos commented:
@tbursztyka You are right sdhc (init priority 40) is initialized before spi (init priority 70) and gpio (init priority 50).

@nzmichaelh It loops in the first do while block in sdhc_detect. I captured an image from my logic analyzer https://gist.githubusercontent.com/ydamigos/c748fe9490ef6f61536dfa7f98118fd2/raw/0df084b41815e21ecbbadb43970db519c1199428/upstream_sdhc_olimexino_spi_signals.png.

ydamigos commented:
>@nzmichaelh It loops in the first do while block in sdhc_detect. I captured an image from my logic analyzer https://gist.githubusercontent.com/ydamigos/c748fe9490ef6f61536dfa7f98118fd2/raw/0df084b41815e21ecbbadb43970db519c1199428/upstream_sdhc_olimexino_spi_signals.png.

Logic analyzer was the culprit. I removed it and now it detects the microSD card. It looks like it detects it two times.

```
**** BOOTING ZEPHYR OS v1.10.99 - BUILD: Mar  2 2018 13:02:39 *****
[sdhc] [INF] sdhc_detect: Found a ~7604 MiB SDHC card.
[sdhc] [INF] sdhc_detect: Manufacturer ID=2 OEM='TM' Name='SA08G' Revision=0x4 Serial=0x9c00014a
[sdhc] [INF] sdhc_detect: Found a ~7604 MiB SDHC card.
[sdhc] [INF] sdhc_detect: Manufacturer ID=2 OEM='TM' Name='SA08G' Revision=0x4 Serial=0x9c00014a

```

nzmichaelh commented:
Hi folks.  The sdhc_init() function (mostly) only sets up bindings so it's safe at any priority.  I've changed it to APPLICATION though so the gpio_configure() happens after the GPIO driver is set up.

@ydamigos sdhc_detect() is called when disk_access_init() is called.  Are you using, say, both the FAT driver and mass storage?  If so it looks like disk_access_init() can get called twice.

ydamigos commented:
>@ydamigos sdhc_detect() is called when disk_access_init() is called. Are you using, say, both the FAT driver and mass storage? If so it looks like disk_access_init() can get called twice.

Yes, I am using mass storage and FAT.

nzmichaelh commented:
I've rebased and fixed the doc comment.

nzmichaelh commented:
@ydamigos I've added duplicate init detection to disk_access_init()

nzmichaelh commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+SDHC disks
```
Done

nzmichaelh commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+SDHC disks
+##########
+
+Zephyr includes support for connecting a SDHC card via the SPI bus.
```
Done

nzmichaelh commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,42 @@
+##########
+
+Zephyr includes support for connecting a SDHC card via the SPI bus.
+When used with ELMFS, this allows reading and writing FAT based
```
ELMFS is used by Zephyr's built-in filesystem support.  Clarified.

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+
+The system has been tested with cards from Samsung, SanDisk, and 4V
+with sizes from 2 GiB to 32 GiB in single partition mode.  Higher
+capacity cards should also work but haven't been tested.  Please let
```
I'd delete, "Please let us know if they work!"

carlescufi commented:
@nzmichaelh can you address @dbkinder's last request and then maybe @tbursztyka will be ready to merge this.

tbursztyka commented on subsys/disk/disk_access_sdhc.c:
```diff
@@ -0,0 +1,981 @@
+static struct sdhc_data sdhc_data_0;
+
+DEVICE_AND_API_INIT(sdhc_0, "sdhc_0", sdhc_init, &sdhc_data_0, NULL,
+		    APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
```
can you change that? APPLICATION level is _never_ for device drivers.

carlescufi commented:
@nzmichaelh @tbursztyka any chance we can rebase this and move it forward?

ramakrishnapallala commented:
Please rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/6970

carlescufi commented:
@nzmichaelh will you be able to rebase this? thanks!

carlescufi commented:
@lemrey @nvlsianpu any chance you could take over this PR?

arturo182 commented:
I am interested in this too, will try to rebase and see how it goes ;)

nashif commented:
Lets close this one and continue in #11793

ntavish commented:
Hi, this does seem to work with the USB mass storage sample, but using the fs shell for example to mount it doesn't work for me. Any idea why that might be?

I am testing this on nrf52840-blip hardware. Using the fs shell, I can issue command `fs mount fat /SDHC` and it shows success. But `fs ls /` or any other operation results in an error.
